internal/graph/check.go

```diff
@@ -0,0 +1,536 @@
+			}
+
+			if !result.resp.GetAllowed() {
+				return result.resp, nil
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+
+	return &openfgapb.CheckResponse{Allowed: true}, nil
+}
+
+// exclusion implements a CheckFuncReducer that requires a 'base' CheckHandlerFunc to resolve to an allowed
+// outcome and a 'sub' CheckHandlerFunc to resolve to a falsey outcome. The base and sub computations are
+// handled concurrently relative to one another.
+func exclusion(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandlerFunc) (*openfgapb.CheckResponse, error) {
+
+	if len(handlers) != 2 {
+		panic(fmt.Sprintf("expected two rewrite operands for exclusion operator, but got '%d'", len(handlers)))
+	}
+
+	limiter := make(chan struct{}, concurrencyLimit)
+
+	cctx, cancel := context.WithCancel(ctx)
+	baseChan := make(chan checkOutcome, 1)
+	subChan := make(chan checkOutcome, 1)
+
+	var wg sync.WaitGroup
+
+	defer func() {
+		cancel()
+		wg.Wait()
+		close(baseChan)
+		close(subChan)
+	}()
+
+	baseHandler := handlers[0]
+	subHandler := handlers[1]
+
+	limiter <- struct{}{}
+	wg.Add(1)
+	go func() {
+		resp, err := baseHandler(cctx)
+		baseChan <- checkOutcome{resp, err}
+		<-limiter
+		wg.Done()
+	}()
+
+	limiter <- struct{}{}
+	wg.Add(1)
+	go func() {
+		resp, err := subHandler(cctx)
+		subChan <- checkOutcome{resp, err}
+		<-limiter
+		wg.Done()
+	}()
+
+	for i := 0; i < len(handlers); i++ {
+		select {
+		case baseResult := <-baseChan:
+			if baseResult.err != nil {
+				return &openfgapb.CheckResponse{Allowed: false}, baseResult.err
+			}
+
+			if !baseResult.resp.Allowed {
+				return &openfgapb.CheckResponse{Allowed: false}, nil
+			}
+		case subResult := <-subChan:
+			if subResult.err != nil {
+				return &openfgapb.CheckResponse{Allowed: false}, subResult.err
+			}
+
+			if subResult.resp.Allowed {
+				return &openfgapb.CheckResponse{Allowed: false}, nil
+			}
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+
+	return &openfgapb.CheckResponse{Allowed: true}, nil
+}
+
+// dispatch dispatches the provided Check request to the CheckDispatcher this ConcurrentChecker
+// was constructed with.
+func (c *ConcurrentChecker) dispatch(ctx context.Context, req *dispatcher.DispatchCheckRequest) CheckHandlerFunc {
+	return func(ctx context.Context) (*openfgapb.CheckResponse, error) {
+		resp, err := c.dispatcher.DispatchCheck(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		return &openfgapb.CheckResponse{
+			Allowed: resp.Allowed,
+		}, nil
+	}
+}
+
+func (c *ConcurrentChecker) DispatchCheck(
+	ctx context.Context,
+	req *dispatcher.DispatchCheckRequest,
+) (*dispatcher.DispatchCheckResponse, error) {
+
+	if req.GetResolutionMetadata().Depth == 0 {
+		return nil, ErrResolutionDepthExceeded
+	}
+
+	typesys, ok := typesystem.TypesystemFromContext(ctx)
+	if !ok {
+		panic("typesystem missing in context")
+	}
+
+	ctx = typesystem.ContextWithTypesystem(ctx, typesys)
```

Comment:
is this line necessary? it's being run in line 288 of `server/server.go`

Reply:
The typesystem is expected to be injected into the `ctx` before the `DispatchCheck` call immediately below. So yeah, this is necessary here.

Reply:
Wait, didn't you just get the typesystem from the context in line 261? Why are you putting it back in?

Reply:
Ah yes, no need. This is superfluous. Will remove.